### PR TITLE
docker: initial checkin Dockerfile for nuttx github action container

### PR DESCRIPTION
**Pull request by liuguo09**
The corresponding docker image:
docker pull liuguo09/ubuntu-nuttx:tagname

Now it is temporarily use personal account to place. Swith to apache docker
once available. Also note the toolchains packages used in Dockerfile are all
locally added which should be downloaded from network instead in future.

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+RUN apt-get install -y curl gettext
+RUN apt-get install -y libcurl4-openssl-dev
+RUN apt-get install -y xxd
+RUN apt-get install -y unzip
+RUN apt-get clean
+RUN mkdir /prebuilt
+
+WORKDIR /prebuilt
+
+ADD gcc-arm-none-eabi.tar.gz ./
```
Comment (xiaoxiang781216) [inline docker/Dockerfile]: change to download from offical release website

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,34 @@
+FROM ubuntu:18.04
```
Comment (xiaoxiang781216) [inline docker/Dockerfile]: Change Dockerfile to another specific name, because other people may add macOS or Windows dockerfile later.

**Comment by btashton** [comment]
@liuguo09 I'll make some detailed comments tonight. Including the multistage image which will resolve the issue of the local builds of the tools.

File: docker/Dockerfile (hunk 1 above)
Comment (liuguo09) [reply docker/Dockerfile]: I'll update it later.

File: docker/Dockerfile (hunk 2 above)
Comment (liuguo09) [reply docker/Dockerfile]: Update it later.

**Comment by liuguo09** [comment]
> @liuguo09 I'll make some detailed comments tonight. Including the multistage image which will resolve the issue of the local builds of the tools.

Wish your feedback : )

**Comment by liuguo09** [comment]
Sorry I don't remember very clearly. It seems in the workflow checkout stages, it mentioned to use 2.18 or later version. With the os version git 2.17.1, it doesn't work well. I'll confirm it later.

**Comment by btashton** [comment]
@liuguo09  Sorry this took me so long.   Have a look at what I have done over here:
https://github.com/btashton/nuttx-build-containers

The docker file is split into a bunch of stages which makes it a lot faster to rebuild and keeps the build time dependencies isolated from the rest of the image.  Note that I am not downloading all of the toolchains now instead of copying local ones in.

I also have a github action wired up to that to create the docker image and place it in the github project package registry. 

On thing I think we should consider is using that structure to create images that each have the needed toolchains they are very large and the ARM builds should not be paying the cost of the RISCV, but I think this is a later improvement that can be made.

You can see this integrated and running here:
https://github.com/btashton/incubator-nuttx/runs/508883969

Let me know how you think and we can look at moving it forward.

**Comment by liuguo09** [comment]
@btashton thanks, I'll look into it and feedback later.

**Comment by liuguo09** [comment]
> @liuguo09 Sorry this took me so long. Have a look at what I have done over here:
> https://github.com/btashton/nuttx-build-containers
> 
> The docker file is split into a bunch of stages which makes it a lot faster to rebuild and keeps the build time dependencies isolated from the rest of the image. Note that I am not downloading all of the toolchains now instead of copying local ones in.

@btashton I notice it reduce half of time for pull the docker. It's really a good job. I totally agree with using a bunch of stages in Dockerfile to faster rebuild.  Do we need a official nuttx build contianer repo or just include it in testing repo? 

> 
> I also have a github action wired up to that to create the docker image and place it in the github project package registry.

Since now we are using our own account to create docker image, how about switch to  the official https://hub.docker.com/u/apache to hold it? I have issued a Jira before  https://issues.apache.org/jira/browse/INFRA-19930 and the Infra team have responded to help setup.  And it required the nuttx committer permission to access. You may the right person to access as you are skilled in docker : )
> 
> On thing I think we should consider is using that structure to create images that each have the needed toolchains they are very large and the ARM builds should not be paying the cost of the RISCV, but I think this is a later improvement that can be made.

If the time reduced not much now, I think it may be well to keep arm and riscv toolchains  together now. But considering more toolchains coming for other archs, it may be reasonable to restructure in later improvement.

> 
> You can see this integrated and running here:
> https://github.com/btashton/incubator-nuttx/runs/508883969
> 
> Let me know how you think and we can look at moving it forward.

**Comment by btashton** [comment]
Did you notice that the docker container is now built via actions and stored in the GitHub package registry? That means it would be stored in the project and not require docker hub. But we can go that way I'm just not sure how much access we have to GitHub secrets for uploading the image.

**Comment by liuguo09** [comment]
Yes, I noticed the docker publish github aciton. It seems more flexible than docker hub.  If taking github package registry approach, do we need an official repo such as incubator-nuttx-build-contianers or use personal contianer project as yours is enough?

**Comment by liuguo09** [comment]
Or is it possible just using the incubator-nuttx-testing to hold dockerfile and do docker publish github action?

**Comment by btashton** [comment]
My plan was to use the incubator-nuttx-testing container. I can submit a PR with this and the action to the repo.

**Comment by liuguo09** [comment]
> My plan was to use the incubator-nuttx-testing repo. I can submit a PR with this and the action to the repo.

That's great!

**Comment by xiaoxiang781216** [comment]
Replace by https://github.com/apache/incubator-nuttx-testing/pull/17, let close it.

**Pull request closed without merging**